### ISSUE TITLE
"Apply discount to Shipping Amount" disabled for "Fixed amount discount for whole cart" action

### ISSUE THREAD

**Issue by drabikowy** [magento/magento2#24422]
### Preconditions (*)
Reproduced on Magento 2.3-develop;

### Steps to reproduce (*)
1. Go to admin panel -> Marketing -> Cart Price Rules -> Add new rule
2. In Actions tab and set Apply to "Fixed amount discount for whole cart"

### Expected result (*)
I am able to turn ON "Apply to Shipping Amount" option in action tab.
It should be possible according to documentation (https://docs.magento.com/m2/ce/user_guide/marketing/price-rules-cart-create.html):

```
Fixed amount discount for whole cart:
Discounts the entire cart by subtracting a fixed amount from the cart total. For example: Enter 10 in Discount Amount to subtract $10 from the cart total.By default, the discount applies only to the cart subtotal. To apply the discount to the subtotal and shipping separately, see Apply to Shipping Amount.
```

### Actual result (*)
"Apply to Shipping Amount" switcher Is disabled

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-19818` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by sudheers-kensium** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @sudheers-kensium. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @sudheers-kensium, here is your [Magento instance](https://i-24422-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-24422-2-3-develop.instances.magento-community.engineering/admin](https://i-24422-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by ghost** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @webkul-deepak. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @webkul-deepak, here is your [Magento instance](https://i-24422-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-24422-2-3-develop.instances.magento-community.engineering/admin](https://i-24422-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by evilprophet** [comment]
I found that it was intentionally removed in version 2.2.9 and 2.3.2
Here is a related commit: https://github.com/magento/magento2/commit/1b6d64c316251604228c32667235b1df9467e0b8
It is a chance to restore this possibility in the future?

**Comment by drabikowy** [comment]
Please give me more details regarding why this specific combination was disabled. 
In mentioned commit there is just a note:
` // Shouldn't be proceed according to MAGETWO-96403` but I believe this is an internal ticket.
I would like to know what issues I can expect when I will restore this option or if it will be fixed with the next release.

**Comment by domi-nique** [comment]
Would be also interesting for us. It is a big problem because we are using this functionality in magento 1. Will it be fixed soon?

**Comment by justinharris1986** [comment]
Throwing my two cents in: This is also causing a problem for us. @magento-engcom-team any idea why this was included? Referencing internal tickets is useless to the community.

**Comment by sivaschenko** [comment]
Issue [MAGETWO-96403](https://jira.corp.magento.com/browse/MAGETWO-96403) was closed as NOT A BUG. Here is the description:

Discount for shipping amount does not applied in case of fixed amount discount for whole cart

**Steps to reproduce:**
- Create product with price - $100
- Create shipping with price - $5
- Create cart rule with:
-- Apply - *Fixed amount discount for whole cart*
-- Discount amount - *1*
--  *Apply to Shipping Amount - Yes*
- Add product to cart, select shipping and check totals

**Actual result:**
```
Subtotal	$100.00
Discount  -$1.00
Shipping (Flat Rate - Fixed)	$5.00
Order Total	$104.00
```

**Expected result:**
Discount should be applied separately for subtotal and shipping (THAT IS NOT CORRECT)
```
Subtotal	$100.00
Discount  -$2.00
Shipping (Flat Rate - Fixed)	$5.00
Order Total	$103.00
```

**The issue is not a bug. "Apply to shipping" is designed for percentage discounts. Fixed discounts are applied correctly.** - the documentation should be updated to clarify this case

**Comment by krzksz** [comment]
@sivaschenko 
> **The issue is not a bug. "Apply to shipping" is designed for percentage discounts. Fixed discounts are applied correctly.** - the documentation should be updated to clarify this case
I'm sorry but I don't understand what are the next steps we can take now, can we revert the changes to restore this feature?

Currently (and it was possible before), there is no way to give coupons to clients which would apply to the whole amount, so the following scenario won't work:
```
Subtotal	$100.00
Coupon  -$150.00
Shipping (Flat Rate - Fixed)	$5.00
Order Total	$0
```

**Comment by drabikowy** [comment]
@sivaschenko Your answer is not clear to me as well. The above example shows that this case should be covered like it was before. 
We tried restoring the code which disabled this possibility but it seems then we have some tax calculation issues (what probably was considered in this commit https://github.com/magento/magento2/commit/1b6d64c316251604228c32667235b1df9467e0b8)

Our real example with this option restored looks like this:
```
Subtotal	10.50
Coupon  -17.56 (I set the cupon value to 20)
Shipping (Flat Rate - Fixed)	7.90
Order Total    0.74
```

**Comment by TheFrakes** [comment]
So does that mean it will be fixed in 2.4 and we can use "apply discount to shipping amount" just like we could in 2.3.1?
![grafik](https://user-images.githubusercontent.com/13464919/72714705-72b38e00-3b6f-11ea-8afb-2518b13cd1fe.png)

**Comment by sivaschenko** [comment]
@krzksz @drabikowy Looks like the internal issue was describing a different scenario - double discount (that was identified as not a bug) - my comment was only sharing the requested internal issue description and decision.

I'd agree that an admin should be able to apply the fixed discount amount to shipping, however, looks like the expectation is that the fixed discount is applied to whole cart (including shipping) - that's why the "Apply to Shipping" switch may not be necessary.

If the fixed discount amount is not applying to shipping - my understanding - that is a bug and it should be fixed.

@tkacheva could you please take a look at this issue and confirm the expected result? Is that a switch and calculation issue or just calculation issue?

**Comment by drabikowy** [comment]
@magento give me 2.3.3 instance

**Comment by magento-engcom-team** [comment]
Hi @drabikowy. Thank you for your request. I'm working on Magento 2.3.3 instance for you

**Comment by magento-engcom-team** [comment]
Hi @drabikowy. 

Thank you for your report and collaboration!

The issue was fixed by Magento team. The fix was delivered into `magento/magento2:2.4-develop` branch(es).
Related commit(s):
 - [45e0641837e88714d279d7b8c175d54edfbf7af2](https://github.com/magento/magento2/commit/45e0641837e88714d279d7b8c175d54edfbf7af2)
 - [21ee858c30fd7ca1bbf0f5be6f0adc45362b774b](https://github.com/magento/magento2/commit/21ee858c30fd7ca1bbf0f5be6f0adc45362b774b)
 - [fbd9c589ac23af18de4f9515421e6fa144277022](https://github.com/magento/magento2/commit/fbd9c589ac23af18de4f9515421e6fa144277022)

The fix will be available with the upcoming `2.4.1` release.
